org.osgi.test.assertj/src/main/java/org/osgi/test/assertj/bundle/AbstractBundleAssert.java

```diff
@@ -0,0 +1,366 @@
+			failWithMessage("%nExpecting%n  <%s>%nto have symbolic name:%n  <%s>%n but was:%n  <%s>", actual,
+				expected, actual.getSymbolicName());
+		}
+		return myself;
+	}
+
+	public SELF hasLocation(String expected) {
+		isNotNull();
+		if (!Objects.equals(actual.getLocation(), expected)) {
+			failWithMessage("%nExpecting%n  <%s>%nto have location:%n  <%s>%n but was:%n  <%s>", actual, expected,
+				actual.getLocation());
+		}
+		return myself;
+	}
+
+	public SELF hasEntry(String entry) {
+		isNotNull();
+		 if (actual.getEntry(entry) == null) {
+			failWithMessage("%nExpecting%n <%s>%nto have entry:%n <%s>%n but it did not", actual, entry);
+		 }
+		return myself;
+	}
+
+	public SELF doesNotHaveEntry(String resource) {
+		isNotNull();
+		URL url = actual.getEntry(resource);
+		if (url != null) {
+			failWithMessage("%nExpecting%n  <%s>%nto not have entry:%n  <%s>%n but it did:%n  <%s>", actual,
+				resource, url);
+		}
+		return myself;
+	}
+
+	public SELF hasResource(String resource) {
+		isNotNull();
+		if (actual.getResource(resource) == null) {
+			failWithMessage("%nExpecting%n  <%s>%nto have resource:%n  <%s>%n but it did not", actual, resource);
+		}
+		return myself;
+	}
+
+	public SELF doesNotHaveResource(String resource) {
+		isNotNull();
+		URL url = actual.getResource(resource);
+		if (url != null) {
+			failWithMessage("%nExpecting%n  <%s>%nto not have resource:%n  <%s>%n but it did:%n  <%s>", actual,
+				resource, url);
+		}
+		return myself;
+	}
+
+	public SELF hasBundleId(long expected) {
+		isNotNull();
+		if (actual.getBundleId() != expected) {
+			failWithMessage("%nExpecting%n  <%s>%nto have bundle ID:%n  <%d>%n but was:%n  <%d>", actual, expected,
+				actual.getBundleId());
+		}
+		return myself;
+	}
+
+	private static Version getVersion(Object version) {
+		if (version == null) {
+			return null;
+		} else if (version instanceof Version) {
+			return (Version) version;
+		} else if (version instanceof CharSequence) {
+			return Version.valueOf(version.toString());
+		} else {
+			throw new IllegalArgumentException(
+				"expected should be an instance of CharSequence or Version; was " + version.getClass()
+					.getName());
+		}
+	}
+
+	public SELF hasVersion(Object expected) {
+		isNotNull();
+		Version expectedVersion = getVersion(expected);
+		if (!Objects.equals(actual.getVersion(), expectedVersion)) {
+			failWithMessage("%nExpecting%n  <%s>%nto have version:%n  <%s>%n but was:%n  <%s>", actual,
+				expected, actual.getVersion());
+		}
+		return myself;
+	}
+
+	public SELF hasVersionGreaterThan(Object expected) {
+		isNotNull();
+		Version expectedVersion = getVersion(expected);
+		if (actual.getVersion()
+			.compareTo(expectedVersion) <= 0) {
+			failWithMessage("%nExpecting%n  <%s>%nto have version greater than:%n  <%s>%n but was:%n  <%s>", actual,
+				expected, actual.getVersion());
+		}
+		return myself;
+	}
+
+	public SELF hasVersionGreaterThanOrEqualTo(Object expected) {
+		isNotNull();
```

Comment:
I like this concept. A couple of hurdles that need to be overcome:

1. It breaks the chain if you're trying to perform multiple asserts against the one bundle;
2. I'm not sure how well the soft assertions proxy will work with this.

I think that both hurdles can be overcome though; I'll have a look at this. It's worth us spending a bit of time now so that we get the code patterns right and re-use them through other assertion classes.

Reply:
I've come up with a pattern that adds an `and()` method to the returned assertion, so that you get the main bundle assertion back again. I'll include it in the next push - 

This fixes the first issue, but the second issue will need a tweak to core AssertJ's soft assertions proxy generation to make it work (the soft proxy generation is expecting the assertion to have a single-arg constructor, but the parent-child chaining requires two args; one to set the parent assertion).